pkg/alertmanager/amcfg.go

```diff
@@ -0,0 +1,514 @@
+// Copyright 2020 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alertmanager
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	monitoringv1beta1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1beta1"
+	"github.com/prometheus/alertmanager/config"
+	commoncfg "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Customization of Config type from alertmanager repo:
+// https://github.com/prometheus/alertmanager/blob/master/config/config.go
+//
+// Custom global type to get around obfuscation of secret values when
+// marshalling. See the following issue for details:
+// https://github.com/prometheus/alertmanager/issues/1985
```

Comment:
I'm not sure this is how we should do it. For Prometheus config, we build the content "manually" using `yaml.MapSlice` and depend on any Prometheus internal data structure.

Reply:
> I'm not sure this is how we should do it.

You don't sound fully convinced by either approach yet, so I'll try to sway you towards this one. :slightly_smiling_face: If I can't make a good case, and there's consensus that using `yaml.MapSlice` is better, then I can try to change it. :+1: 

Is doing everything with `yaml.MapSlice` considered better? Is it more efficient? To me it sounds less nice than unmarshalling the yaml into a type -- less readable and (maybe?) less type safe (`yaml.MapItem` just holding `interface{}` values, etc).

I'm also not sure that this is really depending on an internal alertmanager data structure, since the data structure from the alertmanager project is unfortunately not usable. Instead, this is the prometheus-operator internal type representing the alertmanager config, rather than the alertmanager internal type representing the alertmanager config! :wink:

It's unfortunate that there isn't a type from the Alertmanager project that we can rely on. If there was, and the configuration structure changed in a backwards-incompatible way, then we'd learn at compile-time here, and make any  required changes before releasing a new version of the prometheus-operator that supports that version of Alertmanager.

I manually copied and modified these types from the alertmanager project. That probably could be better...I wonder if it would be good to generate them from the upstream internal types (it's all just a representation of the config yaml that both projects need to understand in the same way)?

Also, these types are _properly_ internal -- we _know_ that nobody will depend on them, unlike in the alertmanager project, where they're _implicitly_ internal, but still exposed, so people could incidentally depend on them. Since they're explicitly an internal implementation detail, we could potentially keep them for now, and if there's a good case for moving to the `yaml.MapSlice` approach in future, we can do that without fear. :slightly_smiling_face:

Reply:
I think having the types is the right way to go, ideally, they will be automatically sync-able in the future. The `yaml.MapSlice` I believe is a major source of allocations and cpu time in the Prometheus config generation, I'd like to one day be able to replace it with a similar mechanism I think (maybe having it typed here is the right way to experiment whether it's the right thing to do).